Branch-SDK/Branch-SDK/BranchUniversalObject.m

```diff
@@ -0,0 +1,137 @@
+//
+//  BranchUniversalObject.m
+//  Branch-TestBed
+//
+//  Created by Derrick Staten on 10/16/15.
+//  Copyright © 2015 Branch Metrics. All rights reserved.
+//
+
+#import "BranchUniversalObject.h"
+#import "BNCError.h"
+#import "BranchConstants.h"
+
+@implementation BranchUniversalObject
+
+- (instancetype)initWithCanonicalIdentifier:(NSString *)canonicalIdentifier {
+    if (self = [super init]) {
+        self.canonicalIdentifier = canonicalIdentifier;
+    }
+    return self;
+}
+
+- (instancetype)initWithTitle:(NSString *)title {
+    if (self = [super init]) {
+        self.title = title;
+    }
+    return self;
+}
+
+- (void)registerView {
+    Branch *branch = [Branch getCurrentInstanceIfAny];
+    if (!branch) {
+        NSLog(@"[Branch Warning] getInstance has not yet been invoked, so could not register view.");
+        return;
+    }
+    if (!self.canonicalIdentifier && !self.title) {
+        NSLog(@"[Branch Warning] a canonicalIdentifier or title are required to uniquely identify content, so could not register view.");
+        return;
+    }
+    
+    [branch registerViewWithParams:[self getParamsForServerRequest] andCallback:nil];
+}
+
+- (void)registerViewWithCallback:(callbackWithParams)callback {
+    Branch *branch = [Branch getCurrentInstanceIfAny];
+    if (!branch) {
+        if (callback) {
+            callback(nil, [NSError errorWithDomain:BNCErrorDomain code:BNCInitError userInfo:@{ NSLocalizedDescriptionKey: @"getInstance has not yet been invoked, so could not register view." }]);
+        }
+        return;
+    }
+    if (!self.canonicalIdentifier && !self.title) {
+        if (callback) {
+            callback(nil, [NSError errorWithDomain:BNCErrorDomain code:BNCInitError userInfo:@{ NSLocalizedDescriptionKey: @"A canonicalIdentifier or title are required to uniquely identify content, so could not register view." }]);
+        }
+        else {
+            NSLog(@"[Branch Warning] a canonicalIdentifier or title are required to uniquely identify content, so could not register view.");
+        }
+        return;
+    }
+    
+    [branch registerViewWithParams:[self getParamsForServerRequest] andCallback:callback];
+}
+
+- (NSString *)getShortUrlWithLinkProperties:(BranchLinkProperties *)linkProperties {
+    Branch *branch = [Branch getCurrentInstanceIfAny];
+    if (!branch) {
+        NSLog(@"[Branch Warning] getInstance has not yet been invoked, so could not generate a URL.");
+        return nil;
+    }
+    if (!self.canonicalIdentifier && !self.title) {
```

Comment:
not sure whether we need this check since BOU can be initialized only  with title or Canonical ID